### ISSUE TITLE
[Distributed] Non-0 ranks creating CUDA contexts on device 0

### ISSUE THREAD

**Issue by kwen2501** [pytorch/pytorch#135279]
### 🐛 Describe the bug

Symptom:
![image](https://github.com/user-attachments/assets/ca41c0f1-9896-47b8-a3d5-962d10c5f71a)

Each non-0 rank is occupying ~ 1GB memory on GPU 0.

### Versions

Simple repro:
```
import torch
import os
import torch.distributed.distributed_c10d as c10d

def repro(rank, world_size):
    device = torch.device("cuda:%d" % rank)
    #torch.cuda.set_device(device)
    c10d.init_process_group(
        backend="nccl", rank=rank, world_size=world_size, device_id=device,
    )
    
    x = torch.ones((10,), device=device)
    c10d.all_reduce(x)
    c10d.destroy_process_group()
    print("clean exit")

if __name__ == "__main__":
    repro(int(os.environ["RANK"]), int(os.environ["WORLD_SIZE"]))

```

Note:
The issue can be avoided if user uncomments the `torch.cuda.set_device(device)` line. But this is not / has not been a requirement of distributed.

**Comment by d4l3k** [comment]
I've always set `CUDA_VISIBLE_DEVICES` instead of torch.cuda.set_device, should we just require/recommend users do that in our tutorials/guides?

**Comment by wconstab** [comment]
CUDA_VISIBLE_DEVICES has an advantage in enforcing the other devices can't be used by accident, but it also prohibits intentional usage of multi-gpu per process, which is a fringe use case that still has some popularity.

I think this is a good thing to align on a separate gh issue- we'd need to update all our docs/tutorials if we want to recommend this.

It still feels like a worthwhile goal to me that we ensure not to activate contexts on devices that aren't the one we initialized pt-d with.  We may want to issue a warning that `init_process_group()` calls without a device argument are deprecated- adding the arg will opt users into 'eager init' for nccl as well as ensure we have a solid target for which device to use for any operations that might be ambiguous.
